euphonic/force_constants.py

```diff
@@ -621,21 +647,49 @@ def _calculate_phonons_at_qpts(
                         q, q_independent_args)
                 if return_eigenvectors:
                     reigenvecs[q] = evecs
-                if return_mode_widths:
+                if return_mode_gradients:
                     rmode_gradients[q] = grads

```

Comment:
Having a little trouble keeping track of the shape of the data through this. Am I right in thinking that
- the einsum in _calculate_phonons_at_q returns a 2D array of modes gradients where the x, y, z components have been serialised along one axis
- these fall back into an x, y, z axis due to the pre-allocated shape of rmode_gradients?

Reply:
Going into the einsum we have 3 arrays:
* `np.conj(evecs_sq_view)`, axes labelled `ij`, shape `(n_modes, n_modes)`
* `evecs_sq_view`, axes labelled `ik`, shape `(n_modes, n_modes)`
* `dmat_grad`, axes labelled `jkl`, shape `(n_modes, n_modes, 3)`

The output from the einsum is:
* `mode_grads_xyz`, axes labelled `il`, shape `(n_modes, 3)`

These output gradients get indexed into `rmode_gradients` (shape `(n_rqpts, n_modes, 3)`) for each q-point. Does that make sense?

Reply:
I think I see now. The `n_modes` on j, k is a combination of atom indices and directions, right? I guess I was confused by the way the cartesian directions in dmat_grad aren't directly connected to the cartesian directions of the atoms. But if that's how the maths works I guess it's ok!

Reply:
Yeah, some codes index it like `(n_atoms, 3, n_atoms, 3)` but I prefer to think of it as a 2D matrix and it makes the einsum stuff easier without having to make different views of the array, but it's just convention I guess